src/OrbitSshQt/TunnelTest.cpp

```diff
@@ -0,0 +1,170 @@
+    QSignalSpy finished_signal{&task_.value(), &orbit_ssh_qt::Task::finished};
+    task_->Start();
+
+    if (!task_->IsStarted()) {
+      QSignalSpy started_signal{&task_.value(), &orbit_ssh_qt::Task::started};
+      EXPECT_TRUE(started_signal.wait());
+    }
+
+    // We have to wait for socat being ready to receive connections. There is no good way to detect
+    // readiness other than parsing the debug output.
+    std::string socat_output{};
+    ScopedConnection output_connection{
+        QObject::connect(&task_.value(), &orbit_ssh_qt::Task::readyReadStdErr, &task_.value(),
+                         [this, &socat_output]() { socat_output.append(task_->ReadStdErr()); })};
+
+    ASSERT_TRUE(QTest::qWaitFor([&socat_output]() {
+      // We need to make sure the first line was printed fully to ensure we can parse the port
+      // number in the next step.
+      return absl::StrContains(socat_output, "listening on AF=2 0.0.0.0:") &&
+             absl::StrContains(socat_output, '\n');
+    })) << "Socat didn't start in time. The debug output was: "
+        << socat_output;
+
+    auto lines = absl::StrSplit(socat_output, '\n');
+    ASSERT_NE(lines.begin(), lines.end());
+    std::string_view first_line = *lines.begin();
+
+    constexpr std::string_view kIpAddress{"0.0.0.0:"};
+    auto ip_location = first_line.find(kIpAddress);
+    ASSERT_NE(ip_location, std::string_view::npos)
+        << "Couldn't find the IP address in the first line: " << first_line;
+
+    std::string_view port_as_string = first_line.substr(ip_location + kIpAddress.size());
+    ASSERT_TRUE(absl::SimpleAtoi(port_as_string, &port_))
+        << "Couldn't parse port number. Input was: " << port_as_string;
+  }
+
+  void TearDown() override {
+    if (task_.has_value()) {
+      KillRunningEchoServer();
+
+      task_->Stop();
+
+      if (!task_->IsStopped()) {
+        QSignalSpy stopped_signal{&task_.value(), &orbit_ssh_qt::Task::stopped};
+        EXPECT_TRUE(stopped_signal.wait());
+      }
+    }
+
+    SshTestFixture::TearDown();
+  }
+
+  [[nodiscard]] uint16_t GetEchoServerPort() const { return static_cast<uint16_t>(port_); }
+
+ private:
+  std::optional<Task> task_;
+  int port_{};
+
+  void KillRunningEchoServer() {
+    // Since we don't have PTY support implemented in `Task`, we can't just send Ctrl+C to the
+    // running process. So instead we execute a kill command in a different task.
+
+    orbit_ssh_qt::Task kill_task{GetSession(),
+                                 absl::StrFormat("kill $(fuser %d/tcp)", GetEchoServerPort())};
+    kill_task.Start();
+
+    if (!kill_task.IsStarted()) {
+      QSignalSpy started_signal{&kill_task, &orbit_ssh_qt::Task::started};
+      EXPECT_TRUE(started_signal.wait());
+    }
+
+    if (!kill_task.IsStopped()) {
+      QSignalSpy stopped_signal{&kill_task, &orbit_ssh_qt::Task::stopped};
+      EXPECT_TRUE(stopped_signal.wait());
+    }
+  }
+};
+
+TEST_F(SshTunnelTest, Start) {
```

Comment:
StartFails? just Start as a name implies it should be successful.

Reply:
Thanks. Done!